mods/pageviews.js

```diff
@@ -0,0 +1,217 @@
+    this.log = options.log || function() {};
+}
+
+
+var tables = {
+        article: 'pageviews.per.article',
+        project: 'pageviews.per.project',
+        tops: 'top.pageviews',
+    };
+var tableURI = function(domain, tableName) {
+        return new URI([domain, 'sys', 'table', tableName, '']);
+    };
+var tableSchemas = {
+        article: {
+            table: tables.article,
+            version: 1,
+            attributes: {
+                project: 'string',
+                article: 'string',
+                access: 'string',
+                agent: 'string',
+                granularity: 'string',
+                // the hourly timestamp will be stored as YYYYMMDDHH
+                timestamp: 'string',
+                views: 'int'
+            },
+            index: [
+                { attribute: 'project', type: 'hash' },
+                { attribute: 'article', type: 'hash' },
+                { attribute: 'access', type: 'hash' },
+                { attribute: 'agent', type: 'hash' },
+                { attribute: 'granularity', type: 'hash' },
+                { attribute: 'timestamp', type: 'range', order: 'asc' },
+            ]
+        },
+        project: {
+            table: tables.project,
+            version: 1,
+            attributes: {
+                project: 'string',
+                access: 'string',
+                agent: 'string',
+                granularity: 'string',
+                // the hourly timestamp will be stored as YYYYMMDDHH
+                timestamp: 'string',
+                views: 'int'
+            },
+            index: [
+                { attribute: 'project', type: 'hash' },
+                { attribute: 'access', type: 'hash' },
+                { attribute: 'agent', type: 'hash' },
+                { attribute: 'granularity', type: 'hash' },
+                { attribute: 'timestamp', type: 'range', order: 'asc' },
+            ]
+        },
+        tops: {
+            table: tables.tops,
+            version: 1,
+            attributes: {
+                project: 'string',
+                access: 'string',
+                year: 'string',
+                month: 'string',
+                day: 'string',
+                // format for this is a json array: [{rank: 1, article: <<title>>, views: 123}, ...]
+                articles: 'string'
+            },
+            index: [
+                { attribute: 'project', type: 'hash' },
+                { attribute: 'access', type: 'hash' },
+                { attribute: 'year', type: 'hash' },
+                { attribute: 'month', type: 'hash' },
+                { attribute: 'day', type: 'hash' },
+            ]
+        }
+    };
+
+/* general handler functions */
+var queryCatcher = function(e) {
+        if (e.status !== 404) {
+            throw e;
+        }
+    };
+var queryResponser = function(res) {
+        // always return at least an empty array so that queries for non-existing data don't error
```

Comment:
I understand the motivation, but, frankly, swallowing errors because of a dashboard's niceties is wrong. All the more so because we are talking about a back-end service/module. We are very far away from the presentation layer here.

IMHO, the separation is quite good if you think about it:

- wrong date or param => receive a 400 (bad request)
- all good, but no data for that period => receive a 200 with an empty set
- no such project/article (resource in general) => receive a 404

> Sadly, there's no way to say "this data should exist but doesn't"

That is partly covered by the second bullet point above, but in general that's the paradigm of a stateless service - a resource is either there or not - there is no previous state form which it can deduce the appropriate *action*. In the REST philosophy it is the client's job to determine the appropriate action to undertake, given that it and only it has the context of the request.